### DISCUSSION TITLE
UQTk version 3.1.5 is now on github

### DISCUSSION THREAD

**Discussion by bjdebus** [sandialabs/UQTk#16] [Announcements]
See https://github.com/sandialabs/UQTk/releases/tag/v3.1.5

**Comment by evasinha-pnnl** [comment]
@bjdebus I upgraded my UQTk to the current 3.1.5 version and then configured and build UQTk again but when running the `ctest` , six of my tests fail: 
```
The following tests FAILED:
	 23 - PyModTest (Failed)
	 27 - PyBCSTest1D (Failed)
	 28 - PyBCSTest2D (Failed)
	 29 - PyBADPTest (Failed)
	 30 - PyRegressionTest (SEGFAULT)
	 31 - PyGalerkinTest (SEGFAULT)
```

Is this something that others are also experiencing?

I also run into `pce_eval: command not found` when using the UQTk tool. Do you have any advice on why the tests are failing and the pce_eval command not found issue?

**Reply by bjdebus** [reply]
Hi @evasinha , sorry you are running into installation issues with the latest UQTk version. Did everything go well during the build process, or did you have any errors?  Sometimes, if the compile goes well, but ctest fails, it could be because your system is not finding python or the PyUQTk modules. Can you check that python is in your path? 
You can also run individual tests by running them from the directory UQTk/build/PyUQTk/pytests. If you run them there directly, they will give you more verbose output.

**Reply by evasinha-pnnl** [reply]
@bjdebus I checked the `build/PyUQTk/pytests/Testing/Temporary/LastTest.log` file and few PyUQTk modules are not being found, as you have suggested.

```
PyUQTk PCE module not found
BCS Regression module not found
Traceback (most recent call last):
  File "/qfs/people/sinh210/build/PyUQTk/pytests/PyBCSTest1D.py", line 101, in <module>
    pcmodel = uqtkpce.PCSet("NISPnoq",nord,ndim,"LEG",0.0,1.0)
NameError: name 'uqtkpce' is not defined
<end of output>
Test time =   1.20 sec 
----------------------------------------------------------
Test Failed.
```

I tested the python version and location as follows:
```
which python
~/.conda/envs/py37/bin/python

python --version
Python 3.9.13
```

What else can I check and fix to ensure all PyUQTk modules can be found?

**Reply by bjdebus** [reply]
Did you check that the modules compiled? Or do you see errors in the build output?

**Reply by evasinha-pnnl** [reply]
Build was complete without any warning or errors.

**Reply by bjdebus** [reply]
Try the following:
export PYTHONPATH="/qfs/people/sinh210/build/PyUQTk/"
if you are using bash or zsh.
Or, if you in csh or tcsh:
setenv PYTHONPATH "/qfs/people/sinh210/build/PyUQTk/"
and then try to run the unit-tests again.
Normally you should not have to include the build folder in your Python Path, but if that works, then I can debug it further from my end.

**Reply by evasinha-pnnl** [reply]
I tried adding `export PYTHONPATH="/qfs/people/sinh210/build/PyUQTk/"` and then running the tests again but the tests are still failing.

**Reply by bjdebus** [reply]
Can you send me the output of the printenv command in your terminal shell?
Can you also send me the output of running "python PyModTest" from directory "/qfs/people/sinh210/build/PyUQTk/pytests"
Thanks!

**Reply by evasinha-pnnl** [reply]
[build.txt](https://github.com/user-attachments/files/18157636/build.txt)
[LastTest.log](https://github.com/user-attachments/files/18157637/LastTest.log)
[printenv.txt](https://github.com/user-attachments/files/18157638/printenv.txt)

@bjdebus I have attached text files showing messages during build, test log, and the printenv command.

I actually don't have `PyModTest.py` in my `/qfs/people/sinh210/build/PyUQTk/pytests/` directory.

Output of two other failed tests are:
```
python PyArrayTest2.py
flags for a_np to show whether C or F contiguous
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False


flags for original numpy array to make sure it hasn't changed to F continguous after converting

Convert uqtk array back to numpy array and make sure C contiguous

compute dot product which should be [2,1.1,6] (Note that if F contigous, the dot product would be [.1,3,6]:
```

```
python PyBCSTest1D.py
/qfs/people/sinh210/build/PyUQTk/pytests
PyUQTk PCE module not found
BCS Regression module not found
Traceback (most recent call last):
  File "/qfs/people/sinh210/build/PyUQTk/pytests/PyBCSTest1D.py", line 101, in <module>
    pcmodel = uqtkpce.PCSet("NISPnoq",nord,ndim,"LEG",0.0,1.0)
NameError: name 'uqtkpce' is not defined
```

**Reply by bjdebus** [reply]
Thanks! I'll take a deeper look, but it may not be till later this afternoon. Thanks for your patience.

**Reply by evasinha-pnnl** [reply]
@bjdebus Thanks for helping me find the cause of the error.

**Reply by bjdebus** [reply]
Thanks for sharing this. When you built UQTk after the upgrade to version 3.1.5, did you start from a clean build directory? The build output looks like it is only compiling files that were updated in the upgrade. 

Can you remove your build directory, make a new one, run the CMake config command again and then build one more time? 

Thanks for continuing to trouble shoot this with me.
Bert

**Reply by evasinha-pnnl** [reply]
@bjdebus I don't remember if I started with a clean build. After reading your response above, I deleted the build directory, made a new one, and ran the cmake config command and make command. The six python tests, however, are still failing, but this time my `build/PyUQTk/pytests` does have `PyModTest.py` and output of `python PyModTest.py` is:

```
Loading array module:
Loading quadrature module:
Loading bayesian compressed sensing module:
Loading polynomial chaos module:
Traceback (most recent call last):
  File "/qfs/people/sinh210/build/PyUQTk/pytests/PyModTest.py", line 48, in <module>
    import _pce
ImportError: /qfs/people/sinh210/build/PyUQTk/pytests/../pce/_pce.cpython-39-x86_64-linux-gnu.so: undefined symbol: N_VSpace
```

**Reply by bjdebus** [reply]
Thanks @evasinha, this is very helpful. I remember running into a similar issue when I was debugging the Sundials upgrade on my Mac. It is possible that when I fixed it, I may not have specified the proper link order of libraries on Linux to resolve this. I will look into this and let you know. I'll try to get this fixed by the end of the week, if not tomorrow.

**Reply by evasinha-pnnl** [reply]
@bjdebus  Thank you very much!

**Reply by bjdebus** [reply]
Hi @evasinha, I was able to replicate the bug you are seeing on one of our linux machines. I am still testing the fix, but for now, can you try the following two fixes:

a) in the file uqtk-snl/UQTk-SNL/PyUQTk/pce/CMakeLists.txt, replace lines 27 - 31 with the following:

```
if(BUILD_SUNDIALS)
	TARGET_LINK_LIBRARIES(_pce PUBLIC uqtk depuqtk blas lapack gfortran ${PROJECT_BINARY_DIR}/../dep/sundials/src/nvector/serial/libsundials_nvecserial.a ${PROJECT_BINARY_DIR}/../dep/sundials/src/cvode/libsundials_cvode.a ${PROJECT_BINARY_DIR}/../dep/sundials/src/sunlinsol/dense/libsundials_sunlinsoldense.a ${PROJECT_BINARY_DIR}/../dep/sundials/src/sunmatrix/dense/libsundials_sunmatrixdense.a ${PROJECT_BINARY_DIR}/../dep/sundials/src/sundials/libsundials_core.a)
else()
	TARGET_LINK_LIBRARIES(_pce PUBLIC uqtk depuqtk blas lapack gfortran ${CMAKE_SUNDIALS_DIR}/lib/libsundials_nvecserial.a ${CMAKE_SUNDIALS_DIR}/lib/libsundials_cvode.a ${CMAKE_SUNDIALS_DIR}/lib/libsundials_sunlinsoldense.a ${CMAKE_SUNDIALS_DIR}/lib/libsundials_sunmatrixdense.a ${CMAKE_SUNDIALS_DIR}/lib/libsundials_core.a)
endif()
```

b) in the file uqtk-snl/UQTk-SNL/PyUQTk/pytests/PyArrayTest2.py, replace np.alltrue with np.all on line 79

Then start from a clean build directory and do a config, make, ctest.

Let me know how it goes. 
Thanks,
Bert

**Reply by evasinha-pnnl** [reply]
@bjdebus, thanks for suggesting fixes. After implementing the fixes all my tests passed :). Thanks for your help in fixing the issue.

**Reply by bjdebus** [reply]
wonderful news. I will push a bug fix to github soon.
